File: Models/baseline_model.R
getwd()


library(reticulate)
library(tidyverse)
library(glmnet)


#py_install('pyarrow')
pandas <- import("pandas")
read_parquet <- function(path, columns = NULL) {
  
  path <- path.expand(path)
  path <- normalizePath(path)
  
  if (!is.null(columns)) columns = as.list(columns)
  
  xdf <- pandas$read_parquet(path, columns = columns)
  
  xdf <- as.data.frame(xdf, stringsAsFactors = FALSE)
  
  dplyr::tbl_df(xdf)
  
}

df = read_parquet('../data/clean/data_journals_merged_topics.gzip')

colnames(df)

# not add topic 3 - since all classified in one of these
independent_var = c('Topic')
dependent_var = c('num_ref')

df_model <- df[,c(dependent_var, independent_var)]
df_model_noNA <- na.omit(df_model)
colnames(df_model_noNA)

df_model_noNA

baseline_topic_model = lm(log(num_ref) ~ as.factor(Topic) , data = df_model_noNA)
summary(baseline_topic_model)




df_word2vec = read.csv('../data/representations/word2vec_doc_representation.csv')
df_word2vec <- df_word2vec %>% select(-X)
df_word2vec <- na.omit(df_word2vec)

x_vars <- model.matrix(num_ref~. , na.omit(df_word2vec))[,-1]
y_var <- log(df_word2vec$num_ref)
lambda_seq <- 10^seq(2, -2, by = -.1)


# Splitting the data into test and train
set.seed(123)
train = sample(1:nrow(x_vars), round(nrow(x_vars)*0.7,0))
x_test = (-train)
y_test = y_var[x_test]


sd(y_test)

cv_output <- cv.glmnet(x_vars[train,], y_var[train],
                       alpha = 1, lambda = lambda_seq, 
                       nfolds = 5)

opt_lambda <- cv_output$lambda.min
# Rebuilding the model with best lamda value identified
lasso_model_word2vec <- glmnet(x_vars[train,], y_var[train], alpha = 1, lambda = opt_lambda)
pred_lasso <- predict(lasso_model_word2vec, s = opt_lambda, newx = x_vars[x_test,])
mean((pred - y_test)^2)

baseline_word2vec_model = lm(y_var[train] ~ . , data = data.frame(x_vars[train,]))
pred_lm = predict(baseline_word2vec_model, newdata =  data.frame(x_vars[x_test,]))
mean((pred_lm - y_test)^2)

File: Models/baseline_model.py
# -*- coding: utf-8 -*-
"""
Created on Sun May 30 10:40:48 2021

@author: flori
"""


import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from helpfunctions_models import regression_results
import statsmodels.api as sm
from scipy import stats



# read with topics 
df = pd.read_parquet('../data/clean/data_journals_merged_topics.gzip')


# get histogram
plt.hist(np.log( 1 + df['num_ref']), bins = 30)

# add logged 
df.columns

independent_var = ['Topic 1', 'Topic 2', 'Topic 3']

model_var = ['num_ref'] + independent_var

df_model = df[model_var]
df_model_noNA = df_model.dropna()

X = df_model_noNA[independent_var]
X2 = sm.add_constant(X)

y_ref = np.log(1 + df_model_noNA['num_ref'])


lm = LinearRegression(fit_intercept=True)
lm.fit(X, y_ref)
y_ref_pred = lm.predict(X)

regression_results(y_ref, y_ref_pred)
lm.coef_